abmarl/policies/q_table_policy.py

```diff
@@ -0,0 +1,156 @@
+
+from abc import ABC, abstractmethod
+
+from gym.spaces import Discrete
+import numpy as np
+
+from .policy import Policy
+
+
+# TODO: Remove this class and let the subclasses, like greedy policy, work with
+# different implementations of the Q insead of forcing it to be a table.
+class QTablePolicy(Policy, ABC):
+    """
+    A policy that explicitly stores and updates a Q-table.
+
+    This requires Discrete observation space and action space.
+    """
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.q_table = np.random.normal(0, 1, size=(self.observation_space.n, self.action_space.n))
+
+    @property
+    def q_table(self):
+        """
+        The Q-table stores the value of each state-action pair.
+
+        The rows of the table should be the observations and the columns should
+        be actions.
+        """
+        return self._q_table
+
+    @q_table.setter
+    def q_table(self, value):
+        assert isinstance(value, np.ndarray), "Q-Table must be a numpy array."
+        assert len(value.shape) == 2, "Q-Table must have two dimensions."
+        assert value.shape[0] == self.observation_space.n, \
+            "The number of rows in the q table must be the same as the size of " + \
+            "the observation space."
+        assert value.shape[1] == self.action_space.n, \
+            "The number of columns in the q table must be the same as the size of " + \
+            "the action space."
+        self._q_table = value
+
+    @property
+    def action_space(self):
+        """
+        The space of possible output actions for this policy.
+
+        Must be Discrete.
+        """
+        return self._action_space
+
+    @action_space.setter
+    def action_space(self, value):
+        assert isinstance(value, Discrete), "Action space must be Discrete."
+        self._action_space = value
+
+    @property
+    def observation_space(self):
+        """
+        The space of possible input observations for this policy.
+
+        Must be Discrete.
+        """
+        return self._observation_space
+
+    @observation_space.setter
+    def observation_space(self, value):
+        assert isinstance(value, Discrete), "Observation space must be Discrete."
+        self._observation_space = value
+
+    @classmethod
+    def build(cls, policy_or_table, **kwargs):
```

Comment:
Not sure that I like the design of policy OR table. It may be better to break this into multiple functions.